crates/cow-amm/src/cache.rs

```diff
@@ -1,24 +1,83 @@
 use {
-    crate::Amm,
+    crate::{Amm, Metrics},
+    anyhow::Context,
     contracts::{CowAmmLegacyHelper, cow_amm_legacy_helper::Event as CowAmmEvent},
+    database::byte_array::ByteArray,
     ethcontract::{Address, errors::ExecutionError},
     ethrpc::block_stream::RangeInclusive,
     shared::event_handling::EventStoring,
-    std::{collections::BTreeMap, sync::Arc},
+    sqlx::PgPool,
+    std::{collections::HashMap, sync::Arc},
     tokio::sync::RwLock,
 };
 
 #[derive(Clone, Debug)]
 pub(crate) struct Storage(Arc<Inner>);
 
 impl Storage {
-    pub(crate) fn new(deployment_block: u64, helper: CowAmmLegacyHelper) -> Self {
-        Self(Arc::new(Inner {
+    pub(crate) async fn new(
+        deployment_block: u64,
+        helper: CowAmmLegacyHelper,
+        factory_address: Address,
+        db: PgPool,
+    ) -> Self {
+        let self_ = Self(Arc::new(Inner {
             cache: Default::default(),
+            factory_address,
             // make sure to start 1 block **before** the deployment to get all the events
             start_of_index: deployment_block - 1,
             helper,
-        }))
+            db,
+        }));
+
+        if let Err(err) = self_.initialize_from_database().await {
+            tracing::error!(
+                ?err,
+                ?factory_address,
+                "failed to initialize AMM cache from database"
+            );
+        }
+
+        self_
+    }
+
+    async fn initialize_from_database(&self) -> anyhow::Result<()> {
+        let mut ex = self.0.db.acquire().await?;
+        let factory_address = ByteArray(self.0.factory_address.0);
+        let db_amms = {
+            let _timer = Metrics::get()
+                .database_queries
+                .with_label_values(&["cow_amm_fetch_by_helper"])
+                .start_timer();
+
+            database::cow_amms::fetch_by_factory_address(&mut ex, &factory_address).await?
+        };
+
+        if db_amms.is_empty() {
+            return Ok(());
+        }
+
+        let mut processed_amms = Vec::new();
+        for db_amm in db_amms {
+            let amm_address = ethcontract::Address::from_slice(&db_amm.address.0);
+            let amm = Amm::new(amm_address, &self.0.helper).await?;
+            let block_number = u64::try_from(db_amm.block_number).context(format!(
+                "db stored cow amm {:?} block number is not u64",
+                db_amm.address
+            ))?;
+            processed_amms.push((block_number, Arc::new(amm)));
+        }
+
+        if !processed_amms.is_empty() {
+            let count = processed_amms.len();
+            let mut cache = self.0.cache.write().await;
+            for (block_number, amm) in processed_amms {
+                cache.entry(block_number).or_default().push(amm);
+            }
+            tracing::info!(count, ?factory_address, "initialized AMMs from database");
```

Comment:
We should probably log all the AMMs that we loaded from the DB so we'll always have at least 1 log stating when the driver became aware of a specific cow amm (e.g. in case it's not getting rebalanced and we want to find out why).